src/commonpp/core/Gelf.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * File: src/commonpp/core/LoggingInterface.cpp
+ * Part of commonpp.
+ *
+ * Distributed under the 2-clause BSD licence (See LICENCE.TXT file at the
+ * project root).
+ *
+ * Copyright (c) 2018 Thomas Sanchez.  All rights reserved.
+ *
+ */
+
+#include "commonpp/core/LoggingInterface.hpp"
+#include "commonpp/core/Utils.hpp"
+#include "commonpp/core/config.hpp"
+#include "commonpp/core/string/json_escape.hpp"
+#include "commonpp/net/Connection.hpp"
+
+#include <boost/asio/io_service.hpp>
+#include <boost/asio/ip/udp.hpp>
+#include <boost/log/attributes/attribute_cast.hpp>
+#include <boost/log/attributes/value_extraction.hpp>
+#include <boost/log/core.hpp>
+#include <boost/log/core/core.hpp>
+#include <boost/log/detail/config.hpp>
+#include <boost/log/expressions.hpp>
+#include <boost/log/expressions/keyword_fwd.hpp>
+#include <boost/log/keywords/severity.hpp>
+#include <boost/log/sinks/basic_sink_backend.hpp>
+#include <boost/log/sinks/frontend_requirements.hpp>
+#include <boost/log/sinks/sync_frontend.hpp>
+#include <boost/log/sinks/text_ostream_backend.hpp>
+#include <boost/log/sources/global_logger_storage.hpp>
+#include <boost/log/sources/record_ostream.hpp>
+#include <boost/log/sources/severity_logger.hpp>
+#include <boost/log/trivial.hpp>
+#include <boost/log/utility/formatting_ostream.hpp>
+#include <boost/log/utility/formatting_ostream_fwd.hpp>
+#include <boost/log/utility/setup/common_attributes.hpp>
+#include <boost/log/utility/setup/formatter_parser.hpp>
+#include <iostream>
+
+#include <iostream>
+#include <random>
+#include <stdint.h>
+#include <string>
+
+namespace logging = boost::log;
+namespace sinks = logging::sinks;
+namespace expr = logging::expressions;
+
+namespace commonpp
+{
+namespace core
+{
+
+class GelfFormatter
+{
+    std::string host_;
+    std::vector<std::pair<std::string, std::string>> static_fields_;
+
+public:
+    GelfFormatter(std::string host,
+                  std::vector<std::pair<std::string, std::string>> static_fields)
+    : host_(std::move(host))
+    , static_fields_(std::move(static_fields))
+    {
+        // it might be possible to "pre-render" these to a fragment of json.
+        for (auto& s : static_fields_)
+        {
+            s.first = commonpp::string::escape_json_string(s.first);
+            s.second = commonpp::string::escape_json_string(s.second);
+        }
+    }
+
+    void operator()(const logging::record_view& rec,
+                    logging::formatting_ostream& strm)
+    {
+        strm << "{";
+        strm << "\"version\" : \"1.1\",";
```

Comment:
shouldn't the version come from the static_fields ? Maybe a list of required fields should be checked ?

Reply:
I don't think so, the version is part of the protocol itself and as a user I don't really want to have to remember to add this tag with a particular version each time. Now I think about it, the static fields need to be prefixed with underscores.

Reply:
works for me